docs/architecture/adr-008-ics20-implementation.md

```diff
@@ -0,0 +1,189 @@
+        /// This function to enable  minting ibc tokens in a module
+        fn mint_coins(&self, module: AccountId, amt: Coin) -> Result<(), ICS20Error>;
+        /// This function should enable burning of minted tokens
+        fn burn_coins(&self, module: AccountId, amt: Coin) -> Result<(), ICS20Error>;
+        /// This function should enable transfer of tokens from the ibc module to an account
+        fn send_coins_from_module_to_account(&self, module: AccountId, to: AccountId, amt: Coin) -> Result<(), ICS20Error>;
+        /// This function should enable transfer of tokens from an account to the ibc module
+        fn send_coins_from_account_to_module(&self, from: AccountId, module: AccountId, amt: Coin) -> Result<(), ICS20Error>;
+    }
+
+    pub trait AccountReader<AccountId: Into<String>> {
+        /// This function should return the account of the ibc module
+        fn get_module_account(&self) -> AccountId;
+        /// Returns the escrow account id for a port and channel combination
+        fn get_escrow_account(port_id: PortId, channel_id: ChannelId) -> AccountId;
+    }
+
+    pub trait ICS20Context: ICS20Keeper + ICS20Reader;
+```
+## Handling ICS20 Packets
+ICS20 messages are still a subset of channel packets so they should be handled as such
+
+The following handlers are recommended to be implemented in the ics20_fungible_token_transfer application in the ibc-rs crate
+
+These handlers will be executed in the module callbacks of any thirdparty IBCmodule that is implementing an ICS20 application on-chain
+```rust
+    /// Initiates ICS20 token transfer
+    /// Escrows tokens from sender and registers the send packet 
+    pub send_transfer<Ctx>(ctx: &Ctx, source_port: PortId, msg: MsgTransfer) -> Result<(), ICS20Error>
+        where Ctx: ICS20Context
+    {
+        /* 
+            ICS20 Application Logic
+            ...
+        */
+    }
+    
+    /// Handles incoming packets with ICS20 data
+    pub on_recv_packet<Ctx>(ctx: &Ctx, packet: Packet, data: MsgTransfer) -> Result<(), ICS20Error>
+        where Ctx: ICS20Context
+    {
+        /* 
+            ICS20 Application Logic
+            ...
+         */
+    }
+
+    /// on_timeout_packet refunds the sender since the original packet sent was
+    /// never received and has been timed out.
+    pub on_timeout_packet<Ctx>(ctx: &Ctx, packet: Packet, data: MsgTransfer) -> Result<(), ICS20Error>
+        where Ctx: ICS20Context
+    {
+        /* 
+            ICS20 Application Logic
+            ...
+        */
+    }
+    /// on_acknowledgement_packet responds to the the success or failure of a packet
+    /// acknowledgement written on the receiving chain. If the acknowledgement
+    /// was a success then nothing occurs. If the acknowledgement failed, then
+    /// the sender is refunded their tokens.
+    pub on_acknowledgement_packet<Ctx>(ctx: &Ctx, packet: Packet, data: MsgTransfer) -> Result<(), ICS20Error>
```

Comment:
IIUC, this function should return an `Acknowledgement` type - both on success and failure. So we're also missing this  `Acknowledgement` type. We'd also want this type to implement `From<GenericAcknowledgement>` or `From<Vec<u8>>`. See [ICS20 technical specification](https://github.com/cosmos/ibc/blob/master/spec/app/ics-020-fungible-token-transfer/README.md#technical-specification) ->
```rust 
pub enum Acknowledgement {
    Success, // => b"AQ=="
    Error(String),
}
```
Additionally, this type must also implement the `Acknowledgement` trait -> https://github.com/informalsystems/ibc-rs/pull/1838#discussion_r809215296